doc/specs/#140 - ZIP Support.md

```diff
@@ -0,0 +1,180 @@
+---
+author: Alexis Bekhdadi @midoriiro
+created on: 2021-11-16
+last updated: 2021-11-17
+issue id: 140
+---
+
+# Zip Support
+
+For [#140](https://github.com/microsoft/winget-cli/issues/140)
+
+## Abstract
+
+This feature adds to winget-cli the ability to install application delivered through a ZIP archive.
+
+## Inspiration
+
+A bunch of applications are only provided by developer's through a ZIP archive. There are three cases:
+1) The archive contains only one binary (a portable / standalone executable).
+2) The archive contains only one supported installer (msix, msi, or exe).
+3) The archive contains more than one file.
+
+## Solution Design
+
+A new installation flow need to be created to execute different installation strategies regarding the three cases above.
+
+First thing first, the archive needs to be extracted in a specific path (TEMP folder can be a good candidate).
+
+Regarding case 1, installation flow check if TEMP folder contain a single binary and repack that binary
+into an MSI(X) package (including setting PATH env var). Then, MSI(X) installation flow can take control.
+
+Regarding case 2, installation flow check if TEMP folder contain an installer (.exe, .msi, .msix) and pass control
+to MSI(X) installation flow.
+
+Regarding case 3, installation flow check if TEMP folder does not contain an installer nor a single binary.
+This case is the complex one, example with 'vagrant' hierarchy:
+- root
+  - bin 
+    - vagrant.exe
+    - other.exe
+  - embedded
+    - bin
+    - some other folders
+
+This hierarchy can be packed into an MSI(X) installer, but how can we specify which binary to add 
+in PATH env var (in that case bin/vagrant.exe)?
+
+This specification take on consideration schema on version 1.1.0. 
+A schema update should be necessary to take all the cases in consideration. 
+
+Take a look into the following manifest examples
+
+### Case 1|2 YAML manifest
+
+```
+PackageIdentifier: Editor.App
+PackageVersion: 1.0
+MinimumOSVersion: 10.0.0.0
+InstallerType: zip
+Scope: ...
+InstallModes: ...
+InstallerSwitches:
+  Silent: ...
+  SilentWithProgress: ...
+UpgradeBehavior: install
+Installers:
+- Architecture: x64
+  InstallerUrl: https://editor.com/app/1.0.zip
+  InstallerSha256: ...
+ManifestType: installer
+ManifestVersion: 1.0.0
+```
+
+This is compliant with the actual schema v1.1.0
+
+### Case 2.1 YAML manifest
+
+If the archive contains an installer but not located at root folder (e.g. archive.zip/somefolder/installer.exe)
+
+```
+PackageIdentifier: Editor.App
+PackageVersion: 1.0
+MinimumOSVersion: 10.0.0.0
+InstallerType: zip
+Scope: ...
+InstallModes: ...
+InstallerSwitches:
+  Silent: ...
+  SilentWithProgress: ...
+UpgradeBehavior: install
+Installers:
+- Architecture: x64
+  InstallerUrl: https://editor.com/app/1.0.zip
+  InstallerSha256: ...
+  InstallerPath: somefolder/installer.exe
```

Comment:
I think we may also need a key defining the secondary installer type.

It is likely that some of the exe’s shipped inside zip packages are actually nullsoft or inno, which should be preferred to exe since the CLI already handles their silent switches without having to define them

Reply:
A key definition already exist on schema [1.1.0](https://github.com/midoriiro/winget-cli/blob/03f8a555693003129e0199759135d874a580a0b5/schemas/JSON/manifests/v1.1.0/manifest.installer.1.1.0.json#L459) (line 459)

Maybe we can use it to tell which installation flow to choose?

Reply:
That key, by definition, would be "zip" since the primary installer is inside of a .zip file. The issue I'm trying to get ahead of is that the zip file could contain different types of exes which have their own installation flow

Reply:
The key I'm pointing could be resolve that issue with something like this:

```
PackageIdentifier: Editor.App
PackageVersion: 1.0
MinimumOSVersion: 10.0.0.0
InstallerType: zip
Scope: ...
InstallModes: ...
InstallerSwitches:
  Silent: ...
  SilentWithProgress: ...
UpgradeBehavior: install
Installers:
- Architecture: x64
  InstallerUrl: https://editor.com/app/1.0.zip
  InstallerSha256: ...
  InstallerPath: somefolder/installer.exe
  InstallerType: inno # or something else
ManifestType: installer
ManifestVersion: 1.0.0
```

Reply:
I don't believe that is valid. Having the key at the manifest level is not compatible with the key at the installer level. One of them will override the other.

Reply:
I see, which component validate the manifest and could be responsible of overriding these key ? An other component or the installation flow ?

Reply:
I'm not sure I understand your question.

The overriding is an inherent property of the CLI. If a key is specified at the installer level, it is overridden by the manifest level. You can't have them defined at both levels
```
PackageIdentifier: ...
PackageVersion: ...
InstallerType: zip <----- THIS IS THE MANIFEST LEVEL
Installers:
- Architecture: ...
  InstallerUrl: ...
  InstallerSha256: ...
  InstallerType: nullsoft<----- THIS IS THE INSTALLER LEVEL
ManifestType: installer
ManifestVersion: 1.0.0
```

Reply:
Ok i see, that's a dead end. About the issue, can you provide a concrete example with multiple installers included in a zip archive ? That could help to better understanding and had a better overview on this issue.

Reply:
I’m not saying multiple installers may be in a .Zip archive. I’m saying a manifest may have different types of installers, and a zip archive may be one of them.

For example - the 7-zip manifest has installers for both exe and msi. Both are in the same manifest.

Some publishers pubish exe’s, msi’s, and zip files, which should all be able to be in a single manifest

Reply:
Ok I understand now. And I agree with you about defining another key to specify the installer type. Maybe InstallerSubType ?

I need to clarify some points on the specification about previous conversations. I'll commit soon.

Reply:
https://github.com/microsoft/winget-cli/issues/1242 is how we're planning to capture some of the "nested installer type" issues. For example, PowerToys is distributed as an .exe, but it installs as an .msi. The installer is a wrapper to help with dependencies.

Reply:
Perfect, I think that as long as that issue is referenced in the spec, then there won't be an issue here

Reply:
That should close this issue. Thank for your feedback.